libs/wingc/src/visit.rs

```diff
@@ -0,0 +1,272 @@
+use crate::ast::{
+	ArgList, Class, Constructor, Expr, ExprKind, FunctionDefinition, InterpolatedStringPart, Literal, Reference, Scope,
+	Stmt, StmtKind,
+};
+
+/// Visitor pattern inspired by implementation from https://docs.rs/syn/latest/syn/visit/index.html
+///
+/// If you are implementing a visitor, you can delegate back to the
+/// default implementation to visit the children of a node by calling
+/// visit::visit_<node_type>(self, node);
+///
+/// For example:
+///
+/// ```ignore
+/// impl<'ast> Visit<'ast> for ExprVisitor {
+///   fn visit_item_fn(&mut self, exp: &'ast Expr) {
+///     println!("Expr with span={}", exp.span);
+///
+///     // Delegate to the default impl to visit any nested expressions.
+///     visit::visit_expr(self, exp);
```

Comment:
IIUC this pattern uses these `visit::visit_X(self, X_struct)` "static" methods as the "default" implementation to achieve some sort of "base class" implementation effect (without inheritance). Would you say that's correct, just so I get what we're doing here?

Reply:
Exactly. To be totally honest before this PR I wasn't even aware this was available, but turns out this is a well supported feature of traits in Rust (even mentioned in the Rust book here: https://doc.rust-lang.org/book/ch10-02-traits.html). 😄